### ISSUE TITLE
coqchk fails on LambdaRust: Anomaly "Uncaught exception Reduction.NotConvertible."

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#6680]
#### Version

git master

#### Operating system
Linux

#### Description of the problem

* Check out https://gitlab.mpi-sws.org/FP/LambdaRust-coq/
* Install the dependencies
* `make -j8 && make validate` fails in coqchk with
```
Error: Anomaly "Uncaught exception Reduction.NotConvertible."
       Please report at http://coq.inria.fr/bugs/.
```

Minimizing this so it "just" depends on Iris results in the following failing file:
```coq
From iris.algebra Require Import csum auth frac gmap agree gset.
From iris.base_logic.lib Require Import own.
Set Default Proof Using "Type".

Module Type lifetime_sig.
End lifetime_sig.

Module temp.
Definition lft_stateR := csumR fracR unitR.

Class lftPreG Σ := LftPreG {
  alft_preG_inG :> inG Σ lft_stateR;
}.

Instance subG_lftPreG Σ :
  lftPreG Σ.
Proof. admit. Admitted.
End temp.


Module Export lifetime : lifetime_sig.
  Include temp.
End lifetime.
```

Cc @ppedrot

**Comment by ppedrot** [comment]
Here is a backtrace on c658141acff6d1b7f610960dd39998385c7e8806:
```
Raised at file "checker/reduction.ml", line 458, characters 11-31
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "checker/reduction.ml", line 137, characters 10-23
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "checker/reduction.ml", line 137, characters 10-23
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "checker/reduction.ml", line 137, characters 10-23
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "clib/cArray.ml", line 436, characters 27-52
Called from file "checker/reduction.ml", line 397, characters 8-43
Called from file "checker/reduction.ml" (inlined), line 549, characters 15-32
Called from file "checker/mod_checking.ml", line 52, characters 4-23
Called from file "list.ml", line 88, characters 24-34
Called from file "checker/mod_checking.ml", line 151, characters 20-111
Called from file "checker/mod_checking.ml", line 92, characters 25-74
Called from file "checker/mod_checking.ml", line 118, characters 15-51
Called from file "list.ml", line 88, characters 24-34
Called from file "checker/mod_checking.ml", line 151, characters 20-111
Called from file "checker/mod_checking.ml", line 89, characters 4-58
Called from file "checker/safe_typing.ml", line 82, characters 2-143
Called from file "checker/check.ml", line 122, characters 6-58
Called from file "list.ml", line 77, characters 12-15
Called from file "checker/check.ml", line 425, characters 2-40
Called from file "checker/checker.ml", line 393, characters 4-20
Error: Anomaly "Uncaught exception Reduction.NotConvertible."
       Please report at http://coq.inria.fr/bugs/.
```

**Comment by RalfJung** [comment]
Here's an entirely dependency-free version:
```coq
Definition gFunctors := nat -> Type.

Class inG (Σ : gFunctors) (A : Type) :=
  InG { inG_id : nat; inG_prf : A = Σ inG_id }.
Arguments inG_id {_ _} _.

Module Type lifetime_sig.
End lifetime_sig.

Module temp.

Class lftPreG Σ := LftPreG {
  alft_preG_inG :> inG Σ nat;
}.

Instance subG_lftPreG Σ :
  lftPreG Σ.
Proof. admit. Admitted.
End temp.


Module Export lifetime : lifetime_sig.
  Include temp.
End lifetime.
```

**Comment by ppedrot** [comment]
Thanks for the minimal example! As expected, this is a nasty case of broken delta-resolving. The conversion problem that fails is
```
(Ind((bug.temp.lftPreG),0,) #1)
≡
(Ind((bug.lifetime.lftPreG,bug.temp.lftPreG),0,) #1)
```
because the checker is unaware that these two things should be the same.

Summoning @maximedenes that claimed to understand how to fix this kind of mess in the checker.

**Comment by maximedenes** [comment]
> @maximedenes that claimed to understand how to fix this kind of mess in the checker.

I don't remember claiming anything like that.

Here's an even more reduced example:

```coq
Module Type S.
End S.

Module M.

Inductive I := C.

Definition t : I.
Proof. exact C. Qed.
  
End M.

Module N : S.
  Include M.
End N.
```

**Comment by maximedenes** [comment]
An interesting experiment would be to remove `Include` from the kernel (by elaborating to aliases).

**Comment by SkySkimmer** [comment]
Fixed by #8773.

**Comment by RalfJung** [comment]
Confirmed. :)
